File: js/stack/20_valid_parentheses.ts
function isValid(s: string): boolean {
  if (s.length <= 1) return false;

  const stack: string[] = [];

  for (let char of s) {
    if (["(", "[", "{"].includes(char)) stack.push(char);

    if (["}", "]", ")"].includes(char)) {
      let last: string = stack.pop() || "";
      if (last === "") return false;
      if (last === "{" && char !== "}") return false;
      if (last === "[" && char !== "]") return false;
      if (last === "(" && char !== ")") return false;
    }
  }

  if (stack.length > 0) return false;

  return true;
}

(() => {
  // Test cases

  // result: false
  console.log(isValid("("));

  // result: true
  console.log(isValid("()"));

  // result: true
  console.log(isValid("()[]{}"));

  // result: false
  console.log(isValid("(])"));

  // result: false
  console.log(isValid(")(){}"));
})();

File: ruby/stack/20_valid_parentheses.rb
# @param {String} s
# @return {Boolean}
def is_valid(s)
  return false if s.size <= 1

  openers = ["(", "[", "{"]
  closers = ["}", "]", ")"]
  opener_to_closer = { "(" => ")", "[" => "]", "{" => "}" }

  stack = []
  s.each_char do |char|
    stack.push char and next if openers.include? char

    next unless closers.include? char

    return false if stack.nil?

    last = stack.pop
    closer = opener_to_closer[last]

    return false unless char.eql? closer
  end

  return false unless stack.empty?

  true
end

def test
  # Test cases

  # result: false
  puts is_valid "("

  # result: true
  puts is_valid "()"

  # result: true
  puts is_valid "()[]{}"

  # result: false
  puts is_valid ")(){}"
end

test
